controller/src/scheduler.rs

```diff
@@ -0,0 +1,430 @@
+use chrono::{DateTime, Utc};
+use cron::Schedule;
+use lazy_static::lazy_static;
+use regex::Regex;
+use snafu::{OptionExt, ResultExt};
+use std::env;
+use std::str::FromStr;
+use tracing::{event, Level};
+use validator::Validate;
+
+// Defines the cron expression scheduler related env variable names
+const SCHEDULER_CRON_EXPRESSION_ENV_VAR: &str = "SCHEDULER_CRON_EXPRESSION";
+
+// Defines the update time window related env variable names
+const UPDATE_WINDOW_START_ENV_VAR: &str = "UPDATE_WINDOW_START";
+const UPDATE_WINDOW_STOP_ENV_VAR: &str = "UPDATE_WINDOW_STOP";
+
+const SCHEDULER_DEFAULT: &str = "* * * * * * *";
+
+/// The module-wide result type.
+type Result<T> = std::result::Result<T, scheduler_error::Error>;
+
+// regex format: HH:MM:SS
+lazy_static! {
+    pub(crate) static ref VALID_UPDATE_TIME_WINDOW_VARIABLE: Regex =
+        Regex::new(r"^((?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d$)").unwrap();
```

Comment:
I think we may have to continue to support the single-digit timestamps here for compatibility reasons.